### PR TITLE
Add a guide to workflow types and fix errors

### PR DESCRIPTION
**Pull request by apmasell**
This adds an explanation for how the Vidarr type system works and fixes some
serialisation/deserialisation inconsistencies.

### CODE REVIEW

File: types.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+type has no meaningful effect; that is optional of optional of string is
+functionally identical to optional of string.
+
+- `{"is": "pair", "left":` _left_`, "right":` _right_ `}`
+A pair of values. This exists mostly as a special case of WDL. It is encoded as
+a JSON object with two properties `"left"` and `"right"`.
+
+- `{"is": "tagged-union", "options": { `_n1_ `:` _t1_`,` ... `} }`
+Create a tagged union type (or algebraic data type). The value provides one
+particular value based on the options available These are encoded as `{"type":
```
Comment (avarsava) [inline types.md:65]: "available These"

File: types.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+A heterogenous tuple of values. A tuple is effectively an object but with
+numerical indices instead of field names. It is encoded as a JSON array.
+
+## Input Types
+All of the conventions for simple types applies to input types. Additionally,
+`"file"` and `"directory"` types are supported.
+
+In both cases, the exact type will depend on the plugins on the server. The submitter can provide either:
+
+- use a file known to Vidarr by its ID: `{"type": "INTERNAL", "contents": ["`_id_`"]}`
```
Comment (avarsava) [inline types.md:83]: "the submitter can provide either use a file [...] or use a file" reads kind of weird

**Review by avarsava — APPROVED** [review]

File: types.md (hunk 2 above)
Comment (apmasell) [reply types.md]: I tried to reword this section.

Comment (avarsava) [reply types.md]: Looks good!

File: types.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+about the contents of a file that can be used by downstream processing. For
+instance, workflow producing a FASTQ or BAM file could include a sequence
+count, so a Shesmu olive consuming this file could pick an appropriate shard
+count for a subsequent workflow.
+
+In some workflows that expand or multiplex content (_e.g_, co-cleaning,
+BCL2FASTQ), the output metadata needs to be dynamically assigned to the output
+files. In that case, the list type is available:
+
+- `{"is": "list", "keys": {`_kn1_`:` _kt1_`,` ... `}, "outputs": {` _on1_`:` _ot1_`, ` ... `}}`
```
Comment (djcooke) [inline types.md:137]: _key-name_, etc.

File: types.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+- `{"is": "optional", "inner":` _type_ `}`
+A value which may be absent. If absent, it is encoded as `null`. Nesting this
+type has no meaningful effect; that is optional of optional of string is
+functionally identical to optional of string.
+
+- `{"is": "pair", "left":` _left_`, "right":` _right_ `}`
+A pair of values. This exists mostly as a special case of WDL. It is encoded as
+a JSON object with two properties `"left"` and `"right"`.
+
+- `{"is": "tagged-union", "options": { `_n1_ `:` _t1_`,` ... `} }`
```
[the remark this answers is not in the document]
Comment (apmasell) [reply types.md]: These are still okay for input and basic types.

**Review by djcooke — APPROVED** [review]

**Pull request merged**